File: ebegu-web/src/utils/DateUtil.ts
/*
 * Ki-Tax: System for the management of external childcare subsidies
 * Copyright (C) 2017 City of Bern Switzerland
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import * as moment from 'moment';
import Moment = moment.Moment; // kann das über ein anderes Import Format gelöst werden (import ... from 'moment')?

export default class DateUtil {

    /**
     * @param {string} localDateTimeString string with format YYYY-MM-DDTHH:mm:ss.SSS
     * @returns {?Moment}
     */
    public static localDateTimeToMoment(localDateTimeString: string): Moment {
        let theMoment: Moment = moment(localDateTimeString, ['YYYY-MM-DDTHH:mm:ss.SSS', 'YYYY-MM-DDTHH:mm:ss', 'YYYY-MM-DDTHH:mm:ss.SSSZ'], true);
        if (!theMoment.isValid()) {
            console.warn('Trying to parse a invalid date to moment', theMoment);
            return undefined;
        }
        return theMoment;
    }

    /**
     * Calls momentToLocalDateFormat with the format by default 'YYYY-MM-DD'
     * @param aMoment
     * @returns {string}
     */
    public static momentToLocalDate(aMoment: Moment): string {
        return DateUtil.momentToLocalDateFormat(aMoment, 'YYYY-MM-DD');
    }

    /**
     * @param {Moment} aMoment time instance
     * @param {string} format the format
     * @returns {?string} a Date (YYYY-MM-DD) representation of the given moment. NULL when aMoment is invalid
     */
    public static momentToLocalDateFormat(aMoment: Moment, format: string): string {
        if (!aMoment) {
            return undefined;
        }
        return moment(aMoment).startOf('day').format(format);
    }

    /**
     * @param {Moment} aMoment time instance
     * @param format format for the time
     * @returns {?string} a Date (YYYY-MM-DD) representation of the given moment. undefined when aMoment is invalid
     */
    public static momentToLocalDateTimeFormat(aMoment: Moment, format: string): string {
        if (!aMoment) {
            return undefined;
        }
        return moment(aMoment).format(format);
    }

    /**
     * @param {Moment} aMoment time instance
     * @returns {?string} a Date (YYYY-MM-DD) representation of the given moment. NULL when aMoment is invalid
     */
    public static momentToLocalDateTime(aMoment: Moment): string {
        return DateUtil.momentToLocalDateTimeFormat(aMoment, 'YYYY-MM-DDTHH:mm:ss.SSS');
    }

    /**
     * @param {string} localDateString string with format YYYY-MM-DD
     * @returns {?moment}
     */
    public static localDateToMoment(localDateString: string): Moment {
        let theMoment = moment(localDateString, 'YYYY-MM-DD', true);
        return theMoment.isValid() ? theMoment : undefined;
    }

    public static today(): Moment {
        return moment().startOf('day');
    }

    public static now(): Moment {
        return moment();
    }

    public static currentYear(): number {
        return moment().year();
    }

    /**
     * Vergleicht 2 Moments. Date und Time werden beruecksichtigt.
     */
    public static compareDateTime(a: Moment, b: Moment): number {
        if (a.isBefore(b)) {
            return -1;
        } else if (a.isSame(b)) {
            return 0;
        } else {
            return 1;
        }
    }
}

File: ebegu-shared/src/main/java/ch/dvbern/ebegu/util/DateUtil.java
/*
 * Ki-Tax: System for the management of external childcare subsidies
 * Copyright (C) 2017 City of Bern Switzerland
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package ch.dvbern.ebegu.util;

import java.sql.Date;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import ch.dvbern.lib.date.feiertage.FeiertageHelper;

import static java.util.Objects.requireNonNull;

/**
 * Utils fuer Date Elemente
 */
public final class DateUtil {

	private DateUtil() {
	}

	/**
	 * Parset den gegebenen String als LocalDate mit dem Format "yyyy-MM-dd"
	 * Sollte der gegebene String null oder leer sein, wird now() zurueckgegeben
	 */
	@Nonnull
	public static LocalDate parseStringToDateOrReturnNow(@Nullable String stringDate) {
		LocalDate date = LocalDate.now();
		if (stringDate != null && !stringDate.isEmpty()) {
			date = parseStringToDate(stringDate);
		}
		return date;
	}

	@Nonnull
	public static LocalDate parseStringToDate(@Nonnull String stringDate) {
		requireNonNull(stringDate);
		return LocalDate.parse(stringDate, DateTimeFormatter.ofPattern("yyyy-MM-dd"));
	}

	/**
	 * Parset den gegebenen String als LocalDateTime mit dem Format "yyyy-MM-dd HH:mm:ss"
	 * Sollte der gegebene String null oder leer sein, wird now() zurueckgegeben
	 */
	@Nonnull
	public static LocalDateTime parseStringToDateTimeOrReturnNow(@Nonnull String stringDateTime) {
		LocalDateTime date = LocalDateTime.now();
		if (!stringDateTime.isEmpty()) {
			date = LocalDateTime.parse(stringDateTime, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
		}
		return date;
	}

	public static boolean isWeekend(@Nonnull LocalDate date) {
		return date.getDayOfWeek() == DayOfWeek.SATURDAY || date.getDayOfWeek() == DayOfWeek.SUNDAY;
	}

	public static boolean isHoliday(@Nonnull LocalDate date) {
		return FeiertageHelper.isFeiertag_CH(Date.valueOf(date));
	}

	public static LocalDate getMax(@Nonnull LocalDate date1, @Nonnull LocalDate date2) {
		return date1.isAfter(date2) ? date1 : date2;
	}

	public static LocalDate getMin(@Nonnull LocalDate date1, @Nonnull LocalDate date2) {
		return date1.isBefore(date2) ? date1 : date2;
	}
}
